Check for items & abilities when parsing damage & heal messages

pmariglia opened the pull request:
Hey Haris :wave: 

Been using poke-env for some personal testing and noticed some information not being parsed from the battle-log.

This PR parses and updates the battle object for

**items** from:
- the item-holder damaging the opponent (example: Rocky Helmet)

**abilities** from:
- the ability healing the pokemon to which it belongs (example: Water Absorb, Volt Absorb)
- the ability damaging the opponent's pokemon (example: Iron Barbs, Rough Skin)

I've tried to write this in a way that I think makes sense with your project - but feel free to criticize anything and everything I wrote :upside_down_face: 

Items healing the pokemon to which they belong is intentionally omitted since they may be consumed during the process (berries), but the message for consumption happens _before_ the -heal message so this would result in an incorrect state. I have a solution for this that I've used in my project but I'll leave that for another discussion/PR.

pmariglia commented on unit_tests/environment/test_battle.py:
```diff
@@ -405,6 +405,56 @@ def test_battle_request_and_interactions(example_request):
+        ['', '-damage', 'p2a: Necrozma', '167/319', '[from] item: Rocky Helmet', '[of] p1a: Groudon']
+    )
+    assert battle.opponent_active_pokemon.item == "rockyhelmet"
+    battle.opponent_active_pokemon._item = None
```
This is one long test, so I reset the item/ability changes like this to make sure a later test doesn't assert something that happened in a previous test. Need to be careful when asserting multiple things in one test :slightly_smiling_face:

hsahovic commented on src/poke_env/environment/abstract_battle.py:
```diff
@@ -210,6 +210,54 @@ def get_pokemon(
+        ):
+            item = split_message[4].split("item:")[-1]
+            pkmn = split_message[5].split("[of]")[-1].strip()
+            self.get_pokemon(pkmn)._item = to_id_str(item)
```
`Pokemon` objects have a setter method associated with the `item` property - could you do

```python
self.get_pokemon(pkmn).item = to_id_str(item)
```
instead? For now it's strictly equivalent, but if might be used for other callbacks at some point.

hsahovic commented on src/poke_env/environment/abstract_battle.py:
```diff
@@ -210,6 +210,54 @@ def get_pokemon(
+        elif len(split_message) == 5 and split_message[4].startswith("[from] item:"):
+            item = split_message[4].split("item:")[-1]
+            pkmn = split_message[2]
+            self.get_pokemon(pkmn)._item = to_id_str(item)
```
Same thing as above :)

hsahovic commented on src/poke_env/environment/abstract_battle.py:
```diff
@@ -210,6 +210,54 @@ def get_pokemon(
+        ):
+            ability = split_message[4].split("ability:")[-1]
+            pkmn = split_message[5].split("[of]")[-1].strip()
+            self.get_pokemon(pkmn)._ability = to_id_str(ability)
```
`ability` also has a setter - I'd change it to `self.get_pokemon(pkmn).ability = to_id_str(ability)`

hsahovic commented on src/poke_env/environment/abstract_battle.py:
```diff
@@ -210,6 +210,54 @@ def get_pokemon(
+        if len(split_message) == 6 and split_message[4].startswith("[from] ability:"):
+            ability = split_message[4].split("ability:")[-1]
+            pkmn = split_message[2]
+            self.get_pokemon(pkmn)._ability = to_id_str(ability)
```
Same as above

hsahovic reviewed (commented):
Thanks a lot for this PR - it's a clear improvement. The only thing that I'd like to change is the setter property things, but it's really minor.
The integration tests are broken because of a typescript change in PS - I'll try to get a fix today, and release it with #176

pmariglia commented:
> Thanks a lot for this PR - it's a clear improvement. The only thing that I'd like to change is the setter property things, but it's really minor.
> The integration tests are broken because of a typescript change in PS - I'll try to get a fix today, and release it with #176

Fixed.

I had an earlier version of the project checked-out and was just copying what was done elsewhere :sweat:. Good to see it's now changed everywhere. :+1:
